### PR TITLE
[4.0] PHPCS Fixes, Abstract Daemon Class

### PR DESCRIPTION
**Pull request by mbabker**
### Summary of Changes
- Fixes the PHPCS issues on the 4.0 branch, includes changing the file comment sniff to ignore the `@package` tag requirement for namespaced code
- Removes the `$config` property declaration in the `Autoconfigurable` app trait; PHP 5.x groans about this being declared in the trait and the class hierarchy
- Makes `JApplicationDaemon` abstract; its parent `JApplicationCli` is also abstract and the class doesn't provide anything useful to keep it concrete
### Testing Instructions

Review the changes
### Documentation Changes Required

Document `JApplicationDaemon` is now abstract.  Odds are though it affects nobody 😉

### CODE REVIEW

File: build/phpcs/Joomla/Sniffs/Commenting/FileCommentSniff.php [hunk 1]
```diff
@@ -362,13 +362,16 @@ protected function processTags($commentStart, $commentEnd)
...
             if ($info['required'] === true && in_array($tag, $foundTags) === false) {
-                $error = 'Missing @%s tag in %s comment';
-                $data  = array(
-                              $tag,
-                              $docBlock,
-                             );
-                $this->currentFile->addError($error, $commentEnd, 'MissingTag', $data);
-                continue;
+                // We don't use package tags in namespaced code or the bootstrap file
+                if ($tag == 'package' && strpos($this->currentFile->getFilename(), '/libraries/src/') === false && strpos($this->currentFile->getFilename(), '/libraries/bootstrap.php') === false) {
```
Comment (zero-24) [inline build/phpcs/Joomla/Sniffs/Commenting/FileCommentSniff.php]: Can we move the opening { please.

File: libraries/legacy/model/admin.php [hunk 2]
```diff
@@ -594,11 +594,14 @@ protected function batchTag($value, $pks, $contexts)
...
 				$tags = array($value);
 
-				$setTagsEvent = Joomla\Cms\Event\AbstractEvent::create('TableSetNewTags', array(
-					'subject'		=> $this,
-					'newTags'		=> $tags,
-					'replaceTags'	=> false
-				));
+				$setTagsEvent = Joomla\Cms\Event\AbstractEvent::create(
+					'TableSetNewTags',
+					array(
```
Comment (zero-24) [inline libraries/legacy/model/admin.php]: I'm not sure is ther any decision made between the old and the new array syntax?

File: tests/unit/suites/libraries/joomla/application/stubs/JApplicationDaemonInspector.php [hunk 3]
```diff
@@ -276,4 +276,17 @@ public function pcntlWait(&$status, $options = 0)
...
+	 * Method to run the application routines.  Most likely you will want to instantiate a controller
+	 * and execute it, or perform some sort of task directly.
+	 *
+	 * @return  void
+	 *
+	 * @since   4.0
+	 */
+	protected function doExecute()
+	{
+		return;
```
Comment (zero-24) [inline tests/unit/suites/libraries/joomla/application/stubs/JApplicationDaemonInspector.php]: Why do we nothing here (i'm sure i just miss something here)
As the doc bock told use we do application rutines instantinate controler etc.?

File: build/phpcs/Joomla/Sniffs/Commenting/FileCommentSniff.php (hunk 1 above)
Comment (mbabker) [inline build/phpcs/Joomla/Sniffs/Commenting/FileCommentSniff.php]: The current PHPCS rule set isn't PHPCS compliant by any measure.  Just leave it be.

File: tests/unit/suites/libraries/joomla/application/stubs/JApplicationDaemonInspector.php (hunk 3 above)
Comment (mbabker) [inline tests/unit/suites/libraries/joomla/application/stubs/JApplicationDaemonInspector.php]: Because the method is declared abstract so it must exist in a concrete implementation.  This is a test stub, it doesn't have to do anything useful.

File: libraries/legacy/model/admin.php (hunk 2 above)
Comment (mbabker) [inline libraries/legacy/model/admin.php]: Considering our verbose code style, I use it more often than not, especially because we enforce a hard line length and shaving 5 characters on a line makes a huge difference.

File: tests/unit/suites/libraries/joomla/application/stubs/JApplicationDaemonInspector.php (hunk 3 above)
Comment (zero-24) [inline tests/unit/suites/libraries/joomla/application/stubs/JApplicationDaemonInspector.php]: Thanks! I have not seen that it was in the tests folder 👍

File: libraries/src/Cms/Application/Autoconfigurable.php [hunk 4]
```diff
@@ -18,14 +16,6 @@
 trait Autoconfigurable
 {
 	/**
-	 * The application configuration object.
-	 *
-	 * @var    Registry
-	 * @since  4.0
-	 */
-	protected $config;
```
Comment (wilsonge) [inline libraries/src/Cms/Application/Autoconfigurable.php]: This seems wrong as the trait hard depends on having a `$config` variable present in the class?

Comment (mbabker) [inline libraries/src/Cms/Application/Autoconfigurable.php]: It does.  But take a look at https://travis-ci.org/joomla/joomla-cms/jobs/160847312#L1137

If we only supported PHP 7 it wouldn't be an issue.  But PHP 5 seems to have an issue with a property being defined this way in both a class and trait.

File: build/phpcs/Joomla/Sniffs/Commenting/FileCommentSniff.php (hunk 1 above)
Comment (photodude) [inline build/phpcs/Joomla/Sniffs/Commenting/FileCommentSniff.php]: Could we make this more general using the T_NAMESPACE token?

Comment (mbabker) [inline build/phpcs/Joomla/Sniffs/Commenting/FileCommentSniff.php]: I was trying to figure that out but is there a `T_NAMESPACE` token available at this point?  It's a file comment sniff, how do you check that at this level?

Comment (photodude) [inline build/phpcs/Joomla/Sniffs/Commenting/FileCommentSniff.php]: Probably need a new member var like `$isNamespaced = false`
which we set to true in the `process()` method if `T_NAMESPACE` exists by checking `$phpcsFile->findNext(T_NAMESPACE, ($stackPtr + 1), null, true);`

(just a guess haven't tested it, could probably make it better)

Comment (photodude) [inline build/phpcs/Joomla/Sniffs/Commenting/FileCommentSniff.php]: or maybe at this point where you added your `if condition` in the processTags
```php
$namespaced = $phpcsFile->findPrevious(T_NAMESPACE, ($stackPtr - 1));
        if ($tag == 'package' && $namespaced !== false) {
```

File: libraries/src/Cms/Application/Autoconfigurable.php (hunk 4 above)
Comment (wilsonge) [inline libraries/src/Cms/Application/Autoconfigurable.php]: Ughhh fair enough. And I assume because it's defined in the framework we can't 'just' omit it from the class?

Comment (mbabker) [inline libraries/src/Cms/Application/Autoconfigurable.php]: Correct.  Unless we're going to start hacking the Composer files or make `Joomla\Application\AbstractApplication` configuration unaware.

File: build/phpcs/Joomla/Sniffs/Commenting/FileCommentSniff.php (hunk 1 above)
Comment (photodude) [inline build/phpcs/Joomla/Sniffs/Commenting/FileCommentSniff.php]: @mbabker this should work correctly (at lease in my tests it seems to)
```php
if ($tag == '@package')
{
    $namespaced = $phpcsFile->findNext(T_NAMESPACE, 0); // this should return 0 if there is no namespaced tokens

    if ($namespaced !== 0) // zero should be correct since it's a count value
    {
        // We don't use package tags in namespaced code
        continue;
    }
}
```

Comment (zero-24) [inline build/phpcs/Joomla/Sniffs/Commenting/FileCommentSniff.php]: @mbabker can we try the suggested changes by @photodude ? If that works we can set this RTC

Comment (photodude) [inline build/phpcs/Joomla/Sniffs/Commenting/FileCommentSniff.php]: BTW I based this change off of the namespace detection in the [PSR2 use Declaration sniff](https://github.com/squizlabs/PHP_CodeSniffer/blob/master/CodeSniffer/Standards/PSR2/Sniffs/Namespaces/UseDeclarationSniff.php#L126-L128)

Comment (mbabker) [inline build/phpcs/Joomla/Sniffs/Commenting/FileCommentSniff.php]: Begrudgingly done.  IMO this is getting more complex than it needs to right now and completely ignored the second part of that conditional where it has a file specific check.  If someone wants to mess with it more do it as a separate PR, let's just get Travis passing here without rewriting the entire PHPCS set.

**Comment by zero-24** [comment]
@mbabker there is just one more left for CS do you want to do it here or in a different PR?

```
FILE: ...d/joomla/joomla-cms/libraries/src/Cms/Session/Storage/JoomlaStorage.php
--------------------------------------------------------------------------------
FOUND 1 ERROR(S) AFFECTING 1 LINE(S)
--------------------------------------------------------------------------------
 56 | ERROR | Arguments with default values must be at the end of the argument
    |       | list
--------------------------------------------------------------------------------
UPGRADE TO PHP_CODESNIFFER 2.0 TO FIX ERRORS AUTOMATICALLY
--------------------------------------------------------------------------------
```

**Comment by mbabker** [comment]
Already a separate PR.  Seems Travis changed something in how it's testing, that file doesn't actually exist in the branch as it's several commits behind the base 4.0 branch yet it's in the report.

**Comment by zero-24** [comment]
RTC. Thanks<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/12067">issues.joomla.org/joomla-cms/12067</a>.</sub>

File: build/phpcs/Joomla/Sniffs/Commenting/FileCommentSniff.php (hunk 1 above)
Comment (photodude) [inline build/phpcs/Joomla/Sniffs/Commenting/FileCommentSniff.php]: Personally, I would rather see the coding standards project for implementing phpcs2 be completed and replace all of the existing cms phpcs set.

Comment (mbabker) [inline build/phpcs/Joomla/Sniffs/Commenting/FileCommentSniff.php]: Ya, the joys of having eleventy seven thousand projects ongoing and a cube root of that number's resources available for it all.

**Comment by mbabker** [comment]
@wilsonge This and https://github.com/joomla/joomla-cms/pull/12141 need to be merged to fix the PHPCS issues, the "false" passing PHPUnit builds on 5.x (the trait composition message), and the heavy memory consumption in PHPUnit.  Either https://github.com/joomla/joomla-cms/pull/12134 needs to be merged or the `JCrypt` constructor fixed to not reference the removed simple crypt class.
